### PR TITLE
Fix MeshTopology geometric dimension

### PR DESCRIPTION
**Pull request by ReubenHill**
Adds the cell geometric dimension to the `ufl_cell` created when creating a `MeshTopology` object. Otherwise the `MeshTopology`'s `ufl_cell` field defaults to the same topological dimension as geometric dimension.

This is necessary to ensure correct creation of `CoordinatelessFunction`s on the mesh topology: UFL uses the UFL cell's geometric dimension information when creating a `CoordinatelessFunction` (rather than looking at the `dim` parameter added when creating the `FunctionSpace` for the `CoordinatelessFunction`). See `ufl.Coefficient.__init__(self, function_space.ufl_element())` on line 62 of `function.py`.

### CODE REVIEW

**Comment by ReubenHill** [comment]
Where should the fix be? Inside `CoordinatelessFunction`? 

In UFL I came across this whilst trying to create a mesh with topologically zero dimensional elements which caused UFL to fail on line `245` of `ufl/domain.py` (function `affine_mesh`). Note that it gets the geometric dimension of the cell on line `243` of `ufl/domain.py`.

**Comment by wence-** [comment]
I'm not sure. But a mesh topology that advertises a cell with a different embedding dimension than topological dimension smells funny.

**Comment by ksagiyam** [comment]
I think there are actually three mesh abstraction levels:
1. topological one (with no concept of geometric dimension)
2. topological one + gdim
3. topological one + coordinates
Type 3 is `MeshGeometry`, which is least flexible. We currently call type 1 `MeshTopology,` which is already enough for `CoordinatelessFunction`. Type 2 can be regarded as type 3 with flexible coordinates; it basically does not distinguish some spherical surface meshes from some ellipsoidal surface meshes.
Type 1 mesh could tell us the minimum geometric dimension it can live in (e.g. by looking at a spherical surface mesh, we can tell that it has to live in three- or higher-dimensional space), but not the exact gdim, so I would like to regard type 2 as different from type1.

I think we need to discuss which abstraction levels (type1, type2 or both) Firedrake wants to have.

**Comment by miklos1** [comment]
Like @wence-, I don't like this idea either.  When you create a `VectorFunctionSpace`, you can specify any dimension you like, although it defaults to the geometric dimension of the underlying mesh, because that's what most people need most of the time.  Rather than leaking geometric information into a `MeshTopology`, the idea was that in case of `gdim != tdim`, you do
```python
coordinate_fs = VectorFunctionSpace(mesh_topology, "CG", 1, dim=gdim)
coordinates = CoordinatelessFunction(coordinate_fs)
```
or something like this. (As in, I'm not sure if the above code is valid Firedrake, but you get the idea...)

**Comment by ReubenHill** [comment]
If `MeshTopology` is actually @ksagiyam 's "type 1" then perhaps the cell geometric dimension should really be `None`? 
I agree that we need to sort out what the abstractions are.

**Comment by miklos1** [comment]
Yes, MeshTopology is type 1 according to @ksagiyam, VectorFunctionSpace on a MeshTopology is more than type 2 but less than type 3, and MeshGeometry is type 3.

I don't mind setting `geometric_dimension`, though I doubt UFL support that, and you may also break assumptions elsewhere in Firedrake.

**Comment by dham** [comment]
I think this probably actually is the right approach. `MeshTopology` is not really type 1, because it contains a Plex, and that Plex knows about geometric dimension. This means that `MeshTopology` is really type 2. It also builds a UFL cell, and those know about geometric dimension too. I think that knowing the correct geometric dimension and not using it just because the class name has "Topology" in it is a perverse thing to do. Remember that we actually make this distinction not because the Topology doesn't have access to information about the geometry, but rather because the class hierarchy would otherwise be circular.

**Comment by wence-** [comment]
> In UFL I came across this whilst trying to create a mesh with topologically zero dimensional elements which caused UFL to fail on line 245 of ufl/domain.py (function affine_mesh). Note that it gets the geometric dimension of the cell on line 243 of ufl/domain.py.

Is the right fix here not to initialise the `ufl.Coefficient` in `CoordinatelessFunction` with an appropriate UFL function space (rather than an element)? I think the PointMeshTopology has a cell with tdim==gdim==0 and a `VectorFunctionSpace(mesh, "DG", 0, dim=0)` "domain"

**Comment by ReubenHill** [comment]
I've pushed some of changes that need to be made to get this passing tests. As I've mentioned on slack, I don't like what this has done to extruded meshes. Their `ufl_cell` can now have two geometric dimensions `TensorProductCell(Cell('interval', 2), interval, geometric_dimension=3)` which seems confusing (though may be a useful descriptor).

**Comment by ReubenHill** [comment]
@wence- That only makes sense if `MeshTopology` is indeed type 1 which @dham and I are not sure it is. For it to be type 1 we would need to remove the plex and be able to tell UFL that the cell's geometric dimension is `None`.

**Comment by wence-** [comment]
Well, the plex is a topological thing on which has been hung geometric information (a plex has a coordinateDM which is like our meshtopology/meshgeometry separation). I don't think we need to remove the plex from a meshtopology. But I think we need to think about what we're going to tell UFL at various points along the way. UFL doesn't know anything concrete about topologies, so we're fine ignoring any plex information. What do we need to inform UFL about. I think, abstractly topological information, which means `tdim == gdim` everywhere in the TopologicalDomain. Then one augments that with some geometric information.

If, on the other hand, UFL doesn't know about topology (and "fakes" it) maybe we should think about adding that abstraction level.

**Comment by wence-** [comment]
A side note, previously we thought 1D was always the hard thing to get right, now it appears you don't even need that many Ds.

File: firedrake/mesh.py [hunk 1]
```diff
@@ -1191,7 +1193,7 @@ def init_cell_orientations(self, expr):
 
         if self.ufl_cell() not in (ufl.Cell('triangle', 3),
                                    ufl.Cell("quadrilateral", 3),
-                                   ufl.TensorProductCell(ufl.Cell('interval'), ufl.Cell('interval'),
+                                   ufl.TensorProductCell(ufl.Cell('interval', geometric_dimension=2), ufl.Cell('interval'),
                                                          geometric_dimension=3)):
```
Comment (miklos1) [inline firedrake/mesh.py:1197]: Yes, this is precisely the thing that, I think, we shall not want to have.

**Comment by ReubenHill** [comment]
- The UFL `AbstractCell` and `AbstractDomain` classes each have `geometric_dimension` and `topological_dimension` properties. 
 - When creating a UFL `Cell(AbstractCell)` the `geometric_dimension == topological_dimension` if `geometric_dimension = None` which it does by default. As far as I can tell this is the only time this occurs.
 - On `AbstractDomain`s, such as `Mesh(AbstractDomain)`, the geometric dimension is taken as given by the `AbstractCell`.

So @wence-  I think that your assumption that, if not specified, `geometric_dimension == topological_dimension` is correct from a UFL point of view. There is, however, no way of making it clear that this is "incomplete" topological-only information, bar it being held in a `MeshTopology`. In which case your solution to specify the UFL domain carefully when creating the CoordinatelessFunction seems reasonable. Ideally it would be possible to guard against this methinks to make sure we aren't making the same mistake elsewhere

**Comment by wence-** [comment]
Proposal: the proximate error is because UFL tries to turn a finite element into a UFL domain (mesh) in `Coefficient.__init__` which we call in `CoordinatelessFunction.__init__`. Instead, we could just build the "right" mesh by hand, something like:

```python
element = ufl.VectorElement("Lagrange", cell, 1, dim=cell.topological_dimension())
mesh = ufl.Mesh(element)
V = ufl.FunctionSpace(mesh, function_space.ufl_element())
ufl.Coefficient.__init__(V)
```

**Comment by wence-** [comment]
Proposal: the proximate error is because UFL tries to turn a finite element into a UFL domain (mesh) in `Coefficient.__init__` which we call in `CoordinatelessFunction.__init__`. Instead, we could just build the "right" mesh by hand, something like:

```python
element = ufl.VectorElement("Lagrange", cell, 1, dim=cell.topological_dimension())
mesh = ufl.Mesh(element)
V = ufl.FunctionSpace(mesh, function_space.ufl_element())
ufl.Coefficient.__init__(V)
```

**Comment by ReubenHill** [comment]
I have created a new PR #1667 with the suggested changes. Closing this.

**Pull request closed without merging**